File: ICPC/April 2026/B.cpp
#include <bits/stdc++.h>
using namespace std;

string transform(const string& num) {

    int zeros = 0;
    for (int i = (int)num.size() - 1; i >= 0 && num[i] == '0'; i--) zeros++;
    if (zeros < 4) return num;
    if (num[0] == '1' && zeros == (int)num.size() - 1) {
        return "10^{" + to_string(zeros) + "}";
    }
    string m = num.substr(0, num.size() - zeros);
    if (m.size() > 1) m.insert(1, ".");
    return m + "\\cdot10^{" + to_string(num.size() - 1) + "}";
    
}

int main() {
    int n;
    cin >> n;
    cin.ignore();
    while (n--) {
        string line;
        getline(cin, line);
        string result;
        vector<bool> stk;
        int i = 0;
        while (i < (int)line.size()) {
            char c = line[i];
            if (c == '{') {
                bool isSub = (i > 0 && (line[i-1] == '^' || line[i-1] == '_'));
                if (!stk.empty() && stk.back()) isSub = true;
                stk.push_back(isSub);
                result += c;
                i++;
            } else if (c == '}') {
                if (!stk.empty()) stk.pop_back();
                result += c;
                i++;
            } else if (isdigit((unsigned char)c)) {
                int start = i;
                while (i < (int)line.size() && isdigit((unsigned char)line[i])) i++;
                string num = line.substr(start, i - start);
                bool insideSub = !stk.empty() && stk.back();
                bool precededSub = (start > 0 && (line[start-1] == '^' || line[start-1] == '_'));
                bool alphaBefore = (start > 0 && isalpha((unsigned char)line[start-1]));
                bool alphaAfter = (i < (int)line.size() && isalpha((unsigned char)line[i]));
                int zeros = 0;
                for (int k = (int)num.size() - 1; k >= 0 && num[k] == '0'; k--) zeros++;
                if (zeros >= 4 && !insideSub && !precededSub && !alphaBefore && !alphaAfter) {
                    result += transform(num);
                } else {
                    result += num;
                }
            } else {
                result += c;
                i++;
            }
        }
        cout << result << "\n";
    }
    return 0;
}

File: ICPC/April 2026/B.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

class B {
    static BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

    static boolean inBraces(String s, int pos) {
        int cnt = 0;
        for (int i = 0; i < pos; i++) {
            if (s.charAt(i) == '{') cnt++;
            else if (s.charAt(i) == '}') cnt--;
        }
        return cnt > 0;
    }

    static String transform(String num) {
        int zeros = 0;
        for (int i = num.length() - 1; i >= 0 && num.charAt(i) == '0'; i--) {
            zeros++;
        }

        if (zeros < 4) return num;

        if (num.charAt(0) == '1' && zeros == num.length() - 1) {
            return "10^{" + zeros + "}";
        }

        String m = num.substring(0, num.length() - zeros);
        if (m.length() > 1) {
            m = m.charAt(0) + "." + m.substring(1);
        }
        return m + "\\cdot10^{" + (num.length() - 1) + "}";
    }

    public static void main(String[] args) throws IOException {
        int n = Integer.parseInt(reader.readLine());

        for (int i = 0; i < n; i++) {
            String line = reader.readLine();
            String result = "";

            for (int j = 0; j < line.length(); j++) {
                if (Character.isDigit(line.charAt(j))) {
                    int start = j;
                    while (j < line.length() && Character.isDigit(line.charAt(j))) {
                        j++;
                    }
                    String num = line.substring(start, j);
                    j--;

                    int zeros = 0;
                    for (int k = num.length() - 1; k >= 0 && num.charAt(k) == '0'; k--) {
                        zeros++;
                    }

                    boolean alphaBefore = (start > 0 && Character.isLetter(line.charAt(start - 1)));
                    boolean alphaAfter = (j + 1 < line.length() && Character.isLetter(line.charAt(j + 1)));

                    if (zeros >= 4 && !inBraces(line, start) && !alphaBefore && !alphaAfter) {
                        result += transform(num);
                    } else {
                        result += num;
                    }
                } else {
                    result += line.charAt(j);
                }
            }

            System.out.println(result);
        }
    }
}
